### ISSUE TITLE
CNF layers very slow with moderately sized Neural Networks

### ISSUE THREAD

**Issue by danharnack** [SciML/DiffEqFlux.jl#413]
Hello, I'm running into a performance issue when using CNF layers for a simple problem. In this code (it's basically the tutorial with a larger network)

```julia
using DiffEqFlux
using OrdinaryDiffEq
using Flux
using Distributions
using Zygote

nn = Chain(Dense(1, 50, tanh),
           Dense(50, 50, tanh),
           Dense(50, 1, tanh))

tspan = (0.0,10.0)
ffjord_test = FFJORD(nn, tspan, Tsit5())

data_train = [Float32(rand(Normal(6.0,0.7))) for i in 1:100]

function loss_adjoint(θ)
    logpx = [ffjord_test(x,θ) for x in data_train]
    loss = -mean(logpx)
end

function train_callback(θ, l)
    println(l)
    false
end

res1 = DiffEqFlux.sciml_train(loss_adjoint,
                              ffjord_test.p,
                              ADAM(0.1),
                              cb = train_callback,
                              maxiters = 10)
```
the first print of the loss happens for me after about 5mins, and it doesn't really speed up after that. I would have expected much faster training times.

I'm using  Julia 1.5.0 and
Zygote v0.5.6
DiffEqFlux v1.22.0
OrdinaryDiffEq v5.42.7
Flux v0.11.1
Distributions v0.23.11

**Comment by ChrisRackauckas** [comment]
I know @avik-pal is playing with this, so I'll see if he has any notes. I just took a look:

https://github.com/SciML/DiffEqFlux.jl/blob/master/src/ffjord.jl#L114-L152

This is correct but this implementation won't scale. It should use the out-of-place form instead of in-place because otherwise it has to use Array{TrackedReal} which explodes the tape size. I'm not sure if that was done to work around an AD issue but it should definitely get fixed because just by looking at the code it's clear it won't scale.

In general, note that in FFJORD implementations right now we're doing a few weird things to work around a bug with Zygote AD nesting. This will get fixed at its source within the next year, but at least in the meantime, we should at least make the tape on arrays instead of scalars.

**Comment by avik-pal** [comment]
Yes indeed these were the stuff I fixed in the Tracker version. https://gist.github.com/avik-pal/746caae4cf04816a8ad7ef17b8c08ce4 has the monte carlo version. There is one bug left to fix which I why I haven't opened a PR yet. It seems to allow backprop only once. In subsequent iterations it errors out saying back! was already used.

Apart from this the Zygote version is also type unstable due to branching. And logpdf promotes things to Float64 which might cause additional slowdown if this layer is used as an intermediate layer.

**Comment by ChrisRackauckas** [comment]
>There is one bug left to fix which I why I haven't opened a PR yet. It seems to allow backprop only once. In subsequent iterations it errors out saying back! was already used.

That's odd. Is this with Tracker on the outside? I see you're using SensitivityADPassThrough

**Comment by avik-pal** [comment]
Yes. It is the outside one. It works on the first iteration and then gives the error if I try to call gradient with new inputs.

**Comment by avik-pal** [comment]
In the meantime, I will open a PR with the Zygote batched version. The example posted also iterates over 100 element array which should also incur some penalty.

On my local machine with a Ryzen 5 CPU I am getting a runtime of about 2.26 mins per epoch.

**Comment by avik-pal** [comment]
On the latest master, the timings for training this model should be

For a Batch size of 100 on CPU, it takes 2.2 mins per iteration of training.
For a Batch size of 1k on GPU, it takes 0.52s per training iteration

**Comment by ChrisRackauckas** [comment]
Great!

**Comment by danharnack** [comment]
Fantastic!! Thank you for the quick fix :)
